finality-aleph/src/new_network/service.rs

```diff
@@ -126,20 +126,26 @@ impl<N: Network, D: Data> Service<N, D> {
         }
     }
 
-    fn next_sender(&mut self) -> Option<NotificationSender> {
+    async fn send(network: &N, send_queue: &mut VecDeque<(D, PeerId, Protocol)>) -> Option<()> {
         loop {
-            let (_, target, protocol) = self.to_send.front()?;
-            if let Some(sender) = self.network.message_sender(*target, protocol.name()) {
-                return Some(sender);
+            if let Some((data, peer, protocol)) = send_queue.front() {
+                if let Ok(()) = network.send(data.encode(), *peer, protocol.name()).await {
+                    send_queue.pop_front();
+                    return Some(());
+                } else {
+                    debug!(target: "aleph-network", "Failed sending data to peer {:?}", peer);
+                }
+                send_queue.pop_front();
```

Comment:
instead of peeking with `front`, the `pop_front` can be used in the if condition and both `pop_front` from branches can be removed

Reply:
Unfortunately we can't do it, as send is asynchronous. It might happen that we pop something, try to send it, another thread finishes first and we loose the data we popped

Reply:
I think I am missing something, in the end, we pop data anyway. What the other thread you have mentioned is doing?

Reply:
But we pop it after the send is successful. This is used in `tokio::select` in `Service::run`.

Reply:
what about the pop on the line 138?

Reply:
Word successful was not a good choice here. After the send is done

Reply:
discussed offline